stream.go

```diff
@@ -0,0 +1,164 @@
+package main
+
+import (
+	"context"
+	"encoding/hex"
+	"fmt"
+	"net/http"
+	"regexp"
+	"strconv"
+	"time"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+	"github.com/google/uuid"
+	"github.com/hashicorp/golang-lru/simplelru"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/rs/zerolog"
+	tmrpcclient "github.com/tendermint/tendermint/rpc/client"
+	rpchttp "github.com/tendermint/tendermint/rpc/client/http"
+	tmjsonclient "github.com/tendermint/tendermint/rpc/jsonrpc/client"
+	tmtypes "github.com/tendermint/tendermint/types"
+)
+
+var (
+	started                = false
+	queryEventBankTransfer = fmt.Sprintf(
+		"%s='%s' AND %s='%s'",
+		tmtypes.EventTypeKey,
+		tmtypes.EventTxValue,
+		"message.action",
+		"/cosmos.bank.v1beta1.MsgSend",
+	)
+	queryInterval = 100 * time.Millisecond
+)
+
+type StreamCollector struct {
+	rpcClient          tmrpcclient.Client
+	logger             zerolog.Logger
+	recentlySeenHashes *simplelru.LRU
+	histogram          *prometheus.HistogramVec
+}
+
+func NewStreamCollector(tmRPC string, logger zerolog.Logger) (*StreamCollector, error) {
+	httpClient, err := tmjsonclient.DefaultHTTPClient(tmRPC)
+	if err != nil {
+		return nil, err
+	}
+
+	httpClient.Timeout = 500 * time.Millisecond
+
+	rpcClient, err := rpchttp.NewWithClient(tmRPC, httpClient)
+	if err != nil {
+		return nil, err
+	}
+	recentlySeen, err := simplelru.NewLRU(10, func(k interface{}, v interface{}) { /* noop */ })
```

Comment:
is `10` the max size for the LRU? Would that be enough if we got a bunch of TXs in one block?

Reply:
nah, the primary reason for this was to intentionally exclude the most recent txs since the event query will return the same latest tx if there isn't a newer one. I'll see how large this reasonably needs to be via testing